Mathlib/Data/Set/Subset.lean

```diff
@@ -0,0 +1,170 @@
+  rw [inter_comm A, inter_comm A, ← Subtype.image_preimage_coe, ← Subtype.image_preimage_coe,
+    image_subset_image_iff Subtype.val_injective]
+
+lemma preimage_val_eq_iff : A ↓∩ B = A ↓∩ C ↔ A ∩ B = A ∩ C := by
+  rw [Subtype.preimage_val_eq_preimage_val_iff, inter_comm _ A, inter_comm _ A]
+
+lemma preimage_val_sUnion : A ↓∩ (⋃₀ S) = ⋃₀ { (A ↓∩ B) | B ∈ S } := by
+  erw [sUnion_image]
+  simp_rw [sUnion_eq_biUnion, preimage_iUnion]
+
+@[simp]
+lemma preimage_val_iInter : A ↓∩ (⋂ i, s i) = ⋂ i, A ↓∩ s i := preimage_iInter
+
+lemma preimage_val_sInter : A ↓∩ (⋂₀ S) = ⋂₀ { (A ↓∩ B) | B ∈ S } := by
+  erw [sInter_image]
+  simp_rw [sInter_eq_biInter, preimage_iInter]
+
+lemma eq_of_preimage_val_eq_of_subset (hB : B ⊆ A) (hC : C ⊆ A) (h : A ↓∩ B = A ↓∩ C) : B = C := by
+  simp only [← inter_eq_right] at hB hC
+  simp only [preimage_val_eq_iff, hB, hC] at h
+  exact h
+
+/-!
+The following simp lemmas try to transform operations in the subtype into operations in the ambient
+type, if possible.
+-/
+
+lemma image_val_univ : ↑(univ : Set A) = A := Subtype.coe_image_univ A
+
+lemma image_val_empty : ↑(∅ : Set A) = (∅ : Set α) := image_empty _
+
+@[simp]
+lemma image_val_union : (↑(D ∪ E) : Set α) = ↑D ∪ ↑E := image_union _ _ _
+
+@[simp]
+lemma image_val_inter : (↑(D ∩ E) : Set α) = ↑D ∩ ↑E := image_inter Subtype.val_injective
+
+@[simp]
+lemma image_val_diff : (↑(D \ E) : Set α) = ↑D \ ↑E := image_diff Subtype.val_injective _ _
+
+@[simp]
+lemma image_val_compl : ↑(Dᶜ) = A \ ↑D := by
+  rw [compl_eq_univ_diff, image_val_diff, image_val_univ]
+
+@[simp]
+lemma image_val_sUnion : ↑(⋃₀ T) = ⋃₀ { (B : Set α) | B ∈ T} := by
+  -- TODO: missing `image_sUnion` lemma
+  erw [sUnion_image]
+  simp_rw [sUnion_eq_biUnion, image_iUnion]
+
+@[simp]
+lemma image_val_iUnion : ↑(⋃ i, t i) = ⋃ i, (t i : Set α) := image_iUnion
+
+@[simp]
+lemma image_val_sInter (hT : T.Nonempty) : (↑(⋂₀ T) : Set α) = ⋂₀ { (↑B : Set α) | B ∈ T } := by
+  erw [sInter_image]
+  rw [sInter_eq_biInter, (Subtype.val_injective.injOn _).image_biInter_eq hT]
+
+@[simp]
+lemma image_val_iInter [Nonempty ι] : (↑(⋂ i, t i) : Set α) = ⋂ i, (↑(t i) : Set α) :=
+  (Subtype.val_injective.injOn _).image_iInter_eq
+
+lemma image_val_subset_set : ↑D ⊆ A := image_val_subset
+
+@[simp]
+lemma image_val_union_self_right_eq : A ∪ ↑D = A :=
+  union_eq_left.2 image_val_subset
+
+@[simp]
+lemma image_val_union_self_left_eq : ↑D ∪ A = A :=
+  union_eq_right.2 image_val_subset
+
+@[simp]
+lemma cou_inter_self_right_eq_coe : A ∩ ↑D = ↑D :=
+  inter_eq_right.2 image_val_subset
+
+@[simp]
+lemma image_val_inter_self_left_eq_coe : ↑D ∩ A = ↑D :=
+  inter_eq_left.2 image_val_subset
+
+@[simp]
+lemma subset_preimage_val_image_val_iff : D ⊆ A ↓∩ ↑E ↔ D ⊆ E := by
+  rw [preimage_image_eq _ Subtype.val_injective]
+
+@[simp]
+lemma image_val_eq_iff : (D : Set α) = ↑E ↔ D = E := Subtype.val_injective.image_injective.eq_iff
+
+lemma image_val_inj (h : (↑D : Set α) = ↑E) : D = E := Subtype.val_injective.image_injective h
```

Comment:
This one should probably be stated via `Function.Injective` and called `image_val_injective`